### PR TITLE
mac,thread: fix pthread_barrier_wait serial thread

### PR DESCRIPTION
**Pull request by ofrobots**
Hi Everyone 👋. This is my first PR here. While trying to fix https://github.com/nodejs/node/issues/23065 using `uv_barriers`, I noticed a thread race on my mac that, I believe, stems from a buggy polyfill of `pthread_barrier_wait` in libuv.

Here's the documentation on `uv_barrier_wait`:
> Note uv_barrier_wait() returns a value > 0 to an arbitrarily chosen “serializer” thread to facilitate cleanup, i.e.
```
if (uv_barrier_wait(&barrier) > 0)
    uv_barrier_destroy(&barrier);
```

The actual implementation returned `PTHREAD_BARRIER_SERIAL_THREAD` (i.e. > 0) on the last thread to enter the wait. IMHO this value should be returned on the last thread to *exit* the wait. It is not safe for any other thread to destroy the barrier as there may be threads yet to exit the barrier.

Checking with implementations elsewhere, refer to the the implementation of [`absl::Barrier::Block`](https://github.com/abseil/abseil-cpp/blob/e01d95528ea2137a4a27a88d1f57c6cb260aafed/absl/synchronization/barrier.cc#L44-L47) from the excellent [abseil](https://abseil.io/) libraries, and they seem to agree with my assessment.

I didn't know about `pthread_barrier_t` until yesterday so it is possible I am figuring things wrong. PTAL.

### CODE REVIEW

**Comment by ofrobots** [comment]
Ping @cjihrig @bnoordhuis @saghul Would you mind taking a peek at this PR?

**Comment by davisjam** [comment]
Hi @ofrobots!

> I noticed a thread race on my mac that, I believe, stems from a buggy polyfill of pthread_barrier_wait in libuv.

Plausible. But from this I understand that you don't have proof, just a guess?

## Is the implementation wrong?

> It is not safe for any other thread to destroy the barrier as there may be threads yet to exit the barrier.

Let us say instead, it is not safe for any other thread to *successfully* destroy the barrier until all threads have exited it safely.

The `pthread_barrier_destroy` docs specify the return of EBUSY for the case:

```
[EBUSY]
    The implementation has detected an attempt to destroy a barrier while it is in use (for example, while being used in a pthread_barrier_wait() call) by another thread.
```

The corresponding libuv polyfill for `pthread_barrier_destroy` includes

```
  if ((rc = pthread_mutex_lock(&b->mutex)) != 0)
    return rc; 

  if (b->in > 0 || b->out > 0)
    rc = EBUSY;
```

As a result, my understanding is that:

1. The polyfill is not incorrect according to the spec, but
2. The example in the libuv docs can fail with `EBUSY`.

## What *should* the implementation do?

The return value `PTHREAD_BARRIER_SERIAL_THREAD` seems useful only from a clean-up perspective.
In this light, the possibility of a subsequent `EBUSY` failure on `pthread_barrier_destroy` adds unnecessary-seeming complexity to application code.
Your patch would address this.
But shouldn't the application code be checking return values no matter what?

**Comment by ofrobots** [comment]
> Plausible. But from this I understand that you don't have proof, just a guess?

Am I reasonable certain? Yes. The crashes that I observe match the thread races as I have described them. With my change, the crashes get fixed.

> The polyfill is not incorrect according to the spec,

I can buy that argument. The more interesting question is whether `uv_barrier_t` has been usefully implemented by libuv. The polyfill is an implementation detail. There are lots of other ways to implement `uv_barrier_t`.

> But shouldn't the application code be checking return values no matter what?

The only documentation for `uv_barrier_t` is the example I pasted above. The example does not work presently on a mac. The EBUSY return value is neither documented, nor does it actually returned on the mac implementation. As you point out, that would be unnecessary complexity for user code anyway.

With that in mind, I stand by the fix I provided here, but I am open to alternative suggestions if you have them.

**Comment by davisjam** [comment]
## Discussion

> The crashes that I observe match the thread races as I have described them.

Maybe I misunderstood the behavior you are experiencing.
1. Are the crashes occurring in one of the `pthread_barrier_X` functions? (if so, we should fix this)
2. Or are they due to an assumption about the state of the barrier in your code? (if so, this is more a question of usability)

> The only documentation for uv_barrier_t is the example I pasted above. 

The libuv docs say [here](http://docs.libuv.org/en/v1.x/threading.html#threading-and-synchronization-utilities) that "The API largely follows the pthreads API", which is why I was looking at the POSIX spec.

## Your patch

I am concerned about your implementation in the case of a barrier initialized with a limit of 1.
It looks like your patch will return 0 in this case?

File: src/unix/thread.c [hunk 1]
```diff
@@ -85,29 +86,30 @@ int pthread_barrier_wait(pthread_barrier_t* barrier) {
...
+     wake up waiters, unlock the mutex. */
   if (++barrier->in == barrier->threshold) {
     barrier->in = 0;
     barrier->out = barrier->threshold - 1;
     rc = pthread_cond_signal(&barrier->cond);
     assert(rc == 0);
 
     pthread_mutex_unlock(&barrier->mutex);
-    return PTHREAD_BARRIER_SERIAL_THREAD;
+    return 0;
```
Comment (davisjam) [inline src/unix/thread.c]: In a 1-thread barrier, shouldn't this return `PTHREAD_BARRIER_SERIAL_THREAD`?

Comment (ofrobots) [reply src/unix/thread.c]: 👍Fixed.

File: src/unix/thread.c [hunk 2]
```diff
@@ -76,6 +76,7 @@ int pthread_barrier_init(pthread_barrier_t* barrier,
 
 int pthread_barrier_wait(pthread_barrier_t* barrier) {
   int rc;
+  char last;
```
[the remark this answers is not in the document]
Comment (ofrobots) [reply src/unix/thread.c]: Fixed.

File: src/unix/thread.c [hunk 3]
```diff
@@ -85,29 +86,30 @@ int pthread_barrier_wait(pthread_barrier_t* barrier) {
...
-    return PTHREAD_BARRIER_SERIAL_THREAD;
+    return 0;
   }
   /* Otherwise, wait for other threads until in is set to 0,
      then return 0 to indicate this is not the first thread. */
   do {
     if ((rc = pthread_cond_wait(&barrier->cond, &barrier->mutex)) != 0)
       break;
   } while (barrier->in != 0);
+  assert(rc == 0);
```
[the remark this answers is not in the document]
Comment (ofrobots) [reply src/unix/thread.c]: Done. There are a few more places where we should be `abort`ing if a pthread function fails illogically. I didn't include those in this PR to keep the review delta small. I think they can be addressed in a follow-on.

File: src/unix/thread.c [hunk 4]
```diff
@@ -85,29 +86,30 @@ int pthread_barrier_wait(pthread_barrier_t* barrier) {
...
 
-  /* mark thread exit */
-  barrier->out--;
+  /* Mark thread exit. If this the last thread to exit, return
+     PTHREAD_BARRIER_SERIAL_THREAD. */
+  last = (--barrier->out == 0);
   pthread_cond_signal(&barrier->cond);
```
[the remark this answers is not in the document]
Comment (ofrobots) [reply src/unix/thread.c]: Done.

**Comment by ofrobots** [comment]
@davisjam 
> Maybe I misunderstood the behavior you are experiencing.
> Are the crashes occurring in one of the pthread_barrier_X functions? (if so, we should fix this)
> Or are they due to an assumption about the state of the barrier in your code? (if so, this is more a question of usability)

I have added a commit with a test (`barrier_serial_thread`) that fails without the fix included in the PR.

File: test/test-barrier.c [hunk 5]
```diff
@@ -139,3 +139,11 @@ TEST_IMPL(barrier_serial_thread) {
   return 0;
 }
 
+/* Single thread uv_barrier_wait should return correct return value. */
+TEST_IMPL(barrier_serial_thread_single) {
+  uv_barrier_t barrier;
+
+  ASSERT(0 == uv_barrier_init(&barrier, 1));
+  ASSERT(0 < uv_barrier_wait(&barrier));
```
Comment (bnoordhuis) [inline test/test-barrier.c]: Can you also call `uv_barrier_destroy()`?

File: test/test-barrier.c [hunk 6]
```diff
@@ -104,3 +104,38 @@ TEST_IMPL(barrier_3) {
 
   return 0;
 }
+
+static void serial_worker(void* data) {
+  uv_barrier_t* barrier = (uv_barrier_t*) data;
```
Comment (bnoordhuis) [inline test/test-barrier.c]: Unnecessary cast.

File: test/test-barrier.c [hunk 7]
```diff
@@ -104,3 +104,38 @@ TEST_IMPL(barrier_3) {
...
+  uv_sleep(100);  /* wait a bit before terminating. */
+}
+
+#define NUM_WORKERS 4
+
+/* Ensure that uv_barrier_wait returns positive only after all threads have
+   exited the barrier. If this value is returned too early and the barrier is
+   destroyed prematurely, then this test may see a crash. */
+TEST_IMPL(barrier_serial_thread) {
+  uv_thread_t threads[NUM_WORKERS];
```
Comment (bnoordhuis) [inline test/test-barrier.c]: It'd be slightly more idiomatic to declare this as `uv_thread_t threads[4];` and then use `ARRAY_SIZE(threads)` everywhere.

File: test/test-barrier.c [hunk 8]
```diff
@@ -104,3 +104,38 @@ TEST_IMPL(barrier_3) {
...
+  uv_barrier_t barrier;
+
+  ASSERT(0 == uv_barrier_init(&barrier, NUM_WORKERS + 1));
+
+  for (int i = 0; i < NUM_WORKERS; ++i) {
```
Comment (bnoordhuis) [inline test/test-barrier.c]: Can you declare `int i` at the top?

File: test/test-barrier.c [hunk 9]
```diff
@@ -104,3 +104,46 @@ TEST_IMPL(barrier_3) {
 
   return 0;
 }
+
+static void serial_worker(void* data) {
+  uv_barrier_t* barrier = (uv_barrier_t*) data;
+  if (uv_barrier_wait(barrier) > 0) {
+    uv_barrier_destroy(barrier);
```
Comment (davisjam) [inline test/test-barrier.c]: Might as well ASSERT this succeeds, here and elsewhere.

**Comment by davisjam** [comment]
> I have added a commit with a test (barrier_serial_thread) that fails without the fix included in the PR.

I have (at last!) identified the cause. Maybe this was obvious to everyone already, but just in case...

I have been studying the source of the libuv polyfills, and I can't find a thread schedule that has `pthread_barrier_wait` and `pthread_barrier_destroy` mis-behaving.

But you're calling `uv_barrier_X` directly! (which makes sense)
The `pthread_barrier_destroy` polyfill returns EBUSY correctly in the event of not-yet-exited threads, but then in `uv_barrier_destroy` we abort on a non-zero rc:

```
void uv_barrier_destroy(uv_barrier_t* barrier) {
  if (pthread_barrier_destroy(barrier))
    abort();
}
```

The `abort` in `uv_barrier_destroy` is incorrect because of the thread schedule you've pointed out, *but it's also incorrect according to the POSIX spec*.
Your PR will cause this abort never to fail, but I don't think the abort should exist at all.

**In summary**:
1. The bug you're reporting could be addressed with a one-liner to remove the `abort` in `uv_barrier_destroy`. Unfortunately, `uv_barrier_destroy` returns void so there would be no way to propagate the error. Users could leak memory in the event of an error.
2. Your patch takes the approach of preventing the "bad" schedule from occurring, so there will be no memory leaks when using the libuv polyfill. But this non-spec-compliant `abort` may still cause program crashes in other implementations of pthread.

I don't like the prospect of memory leaks, so I support your patch.
I also don't like the idea of non-spec-compliant aborts, so I think we should remove this abort as part of the PR.

File: test/test-barrier.c (hunk 5 above)
Comment (ofrobots) [reply test/test-barrier.c]: Done.

File: test/test-barrier.c (hunk 6 above)
Comment (ofrobots) [reply test/test-barrier.c]: Fixed.

File: test/test-barrier.c (hunk 7 above)
Comment (ofrobots) [reply test/test-barrier.c]: Done.

File: test/test-barrier.c (hunk 8 above)
Comment (ofrobots) [reply test/test-barrier.c]: Done.

File: test/test-barrier.c (hunk 9 above)
Comment (davisjam) [reply test/test-barrier.c]: My mistake. I posted this before noticing that returned void. I assumed it was just a pass-through to `pthread_barrier_destroy` with the same signature.

**Comment by ofrobots** [comment]
@davisjam 
> I also don't like the idea of non-spec-compliant aborts, so I think we should remove this abort as part of the PR.

It is not clear (to me) which spec applies here anyway. `uv_barrier_t` have very little documentation. There are hints that for thread APIs in general that they are supposed to "largely follow" pthread, but the barrier docs and the code sample provided don't really seem to be mirroring pthread at this point. For example, the `void` return on `uv_barrier_wait`.

I suggest we move this discussion to another issue as it is not necessary to resolve the bug that I am trying to fix.

@bnoordhuis I've squashed the commit and address the rest of the comments.

Thank you both for your review and insights!

**Comment by cjihrig** [comment]
Should this be targeting v1.x?

**Comment by davisjam** [comment]
Agreed, @ofrobots can you rebase?

**Comment by ofrobots** [comment]
@cjihrig @davisjam Done.

**Comment by ofrobots** [comment]
@davisjam can we fork that discussion into a separate issue? I would this change to land. This fix is blocking a cascade of PRs in Node.js.

**Comment by davisjam** [comment]
@ofrobots Sure. It's not a blocker for me.

**Comment by ofrobots** [comment]
What are the next steps here, is there a CI somewhere that can be run?

File: src/unix/thread.c [hunk 10]
```diff
@@ -83,7 +83,7 @@ int pthread_barrier_init(pthread_barrier_t* barrier,
 }
 
 int pthread_barrier_wait(pthread_barrier_t* barrier) {
-  int rc;
+  int rc, last;
```
Comment (bnoordhuis) [inline src/unix/thread.c]: Minor style issue: one declaration per line.

Comment (ofrobots) [reply src/unix/thread.c]: Done.

**Comment by ofrobots** [comment]
Windows test failure seems unrelated. AIX failed is relevant and needs to be investigated. I have opened https://github.com/nodejs/build/issues/1514 to get login access to AIX to be able to investigate this.

**Comment by ofrobots** [comment]
It seems that the AIX implementation of `pthread_barrier_wait` does indeed return `PTHREAD_BARRIER_SERIAL_THREAD` early – before the other threads have exited the barrier.

This means that that default example of `uv_barrier_wait` in docs may abort as @davisjam was worried about:

```c++
if (uv_barrier_wait(&barrier) > 0)
    uv_barrier_destroy(&barrier);
```

I am forced to conclude that the `abort()` in `uv_barrier_destroy()` is not safe on AIX, and that we should change the default example to something like this instead:

```
uv_barrier_wait(&barrier);
if (0 == uv_barrier_destory(&barrier)) {
  // any additional cleanup, e.g. free of the barrier.
}
```

**Comment by ofrobots** [comment]
I also noticed that we have AIX 6.1 in the CI. As per https://www-01.ibm.com/support/docview.wss?uid=isg3T1012517, AIX 6.1 may already be EOL since last year? It is possible that the behaviour is different/fixed on AIX 7.

Looking at node supported platforms: https://github.com/nodejs/node/blob/master/BUILDING.md#supported-platforms-1, AIX 7 is the minimum supported. libuv claims to [support](https://github.com/libuv/libuv/blob/v1.x/SUPPORTED_PLATFORMS.md) AIX 6 though.

**Comment by bnoordhuis** [comment]
Maybe we should use our own barrier implementation everywhere for the sake of uniform behavior, or at least switch to it on AIX. I've opened #2019 and included the changes from this PR.

**Comment by ofrobots** [comment]
👍. Deferring to #2109.
